### PR TITLE
fix irrigation schedule

### LINKED ISSUE #1482: Error in the Irrigation module
Hello,

I have noticed that the new irrigation module incorporated in the last WRF version 4.2 does not work as it is expected. The irrigation schedule is not properly defined. There is irrigation at all hours, not between the specified times ( irr_start_hour                      and  irr_start_hour + irr_num_hours) defined in the namelist.input. This is observed in the module phys/module_sf_irrigation.F in the IF conditional of the three subroutines (drip_irrigation, channel_irrigation and sprinkler_irrigation):

   xt24=mod(xtime,1440.)
   tloc=floor(gmt+xt24/60.)
   if(tloc.lt.0) tloc=tloc+24
   IF (((tloc.GE.irr_start_hour .AND. tloc.LT.24) .OR. ( tloc.GE.0 .AND. tloc.LT.end_hour) ) &
               .AND. ((julian_in.GE.irr_start_julianday  .AND.  julian_in.LT.367) .OR.       &
             (  julian_in.GE.0 .AND.   julian_in.LT.irr_end_julianday)) .AND. timing.EQ.0.  ) THEN
     IF(irr_ph.EQ.0) THEN
...

It should be:

   xt24=mod(xtime,1440.)
   tloc=floor(gmt+xt24/60.)
   if(tloc.lt.0) tloc=tloc+24
   IF (((tloc.GE.irr_start_hour .AND. tloc.LT.24) .AND. ( tloc.GE.0 .AND. tloc.LT.end_hour) ) &
               .AND. ((julian_in.GE.irr_start_julianday  .AND.  julian_in.LT.367) .AND.       &
             (  julian_in.GE.0 .AND.   julian_in.LT.irr_end_julianday)) .AND. timing.EQ.0.  ) THEN
     IF(irr_ph.EQ.0) THEN
...

Otherwise it will irrigate always. 

Greetings,

Ricard

### PR DESCRIPTION
**Pull request by arjanna**
TYPE: bug fix

KEYWORDS: irrigation schedule, fix

SOURCE: internal

DESCRIPTION OF CHANGES:
Problem:
wrong day-span selection. There were problems when either with Julian day=0 or without.

Solution:
new function to select the irrigation period

ISSUE: For use when this PR closes an issue.
Fixes #1482

LIST OF MODIFIED FILES: 
M	phys/module_irrigation.F

TESTS CONDUCTED: 
1. Do mods fix problem? I checked the compilation. Need further testing with data.
2. Jenkins tests are all passing.

RELEASE NOTE: Fix for the irrigation scheduling issue.

### CODE REVIEW

**Comment by dudhia** [comment]
Keywords should specify which irrigation flag this is as we will have another one in V4.3

**Comment by arjanna** [comment]
If I found the issue correctly, it is the same problem as the previous one... This fix should be ok now. 

Can some test it for periods that both include the new year and not? You should see that now it works for both.

**Comment by arjanna** [comment]
@davegill could you give me the log of the error please?

**Comment by davegill** [comment]
@arjanna 
I just re-submitted the jenkins test. I think that you got a false failure.

**Comment by davegill** [comment]
jenkins
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: b21ecf4c376a4d4b2eab4d4637a7ebc6f8cf5ce7, requested by: arjanna for PR: https://github.com/wrf-model/WRF/pull/1492. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 163           161        0
    Number of Comparisons  : 103           102        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None
```

**Comment by arjanna** [comment]
@davegill thanks! 
@rsegura24 could you check it works with the fix?

thanks

**Comment by dudhia** [comment]
I will approve it. If any more updated testing can be added to the comment, it should be.

**Review by dudhia — APPROVED** [review]

**Comment by rsegura24** [comment]
Hi @arjanna,

Could you tell me where this new version fixed is?

Thanks

Ricard

**Comment by rsegura24** [comment]
Hi @arjanna,

I think that I have found the new version. I believe that it is still wrong if it is the version that I have found in arjanna:irrigation_schedule

IF (((tloc.GE.irr_start_hour .AND. tloc.LT.24) **.OR.** ( tloc.GE.0 .AND. tloc.LT.end_hour) ) &
               .AND. irr_on.eq.1 .AND. timing.EQ.0.  ) THEN
...

The .OR. conditional  I have bolded makes it irrigate during all the hours. Not between irr_start_hour and end_hour. 
For example: if irr_start_hour = 6, end_hour = 8 and tloc = 9 the module should not be activated. However, the first conditional (tloc.GE.irr_start_hour .AND. tloc.LT.24) will be True and the second conditional  ( tloc.GE.0 .AND. tloc.LT.end_hour) False, but the union of both conditionals is still True due to the .OR. operator. I hope it makes it clearer now what I was pointing.

By the way, great work with this irrigation module. It is very simple and useful. Congratulations!

Greetings,

Ricard

**Comment by dudhia** [comment]
I think I agree with the above comment.
The logic should check tloc between start and end for the normal case (end
> start)
then for the through 0 hour case (end < start) would have a different logic
like tloc not between end and start.
Jimy

**Review by dudhia — CHANGES_REQUESTED** [review]
logic fix required before approavl

**Comment by arjanna** [comment]
I will check it. I will do it at the end of this week or the beginning of the next week. I aim to implement it in a way that the selection function irrigation_on will have all the time selection (both day and hour).

@rsegura24 thanks :D

**Comment by dudhia** [comment]
@arjanna can you try to do this quickly. The 4.3 release is about ready to go.

**Comment by arjanna** [comment]
I committed the fixes, please have a look. If possible trigger Jenkins. 

@rsegura24 please also have a look if they works with your case study.

p.s. I could not test them within WRF becuase an issue I reported.

File: phys/module_irrigation.F [hunk 1]
```diff
@@ -1,6 +1,46 @@
...
+   END IF
+ END IF 
+ 
+ IF(irr_start_hour .lt. end_hour) THEN
+   IF((tloc.GE.irr_start_hour .AND. tloc.LT.end_hour))  THEN
+      irr_on_time=1
+   END IF
+ ELSE
+   IF((tloc.GE.irr_start_hour  .AND.  tloc.LT.24) .OR.  &
+     (  tloc.GE.0 .AND.   tloc.LT.end_hour))  THEN
```
Comment (dudhia) [inline phys/module_irrigation.F:36]: This could have been simpler logic if it was just .not. between start and end.

Comment (arjanna) [reply phys/module_irrigation.F:36]: yes. I change it now

**Comment by davegill** [comment]
jenkins
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 3b2550e31e715fd0cbf8c29539b2df070a68e0ad, requested by: arjanna for PR: https://github.com/wrf-model/WRF/pull/1492. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 163           161        0
    Number of Comparisons  : 103           102        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None
```

**Comment by arjanna** [comment]
yes, thanks! Today I got the report from Jenkins

**Comment by dudhia** [comment]
Thanks, yes, the change looks good.

**Review by dudhia — APPROVED** [review]

**Review by dudhia — APPROVED** [review]

**Comment by weiwangncar** [comment]
@rsegura24 Any chance of doing a test for the latest change?

**Comment by weiwangncar** [comment]
Are we willing to accept this PR without confirmation that this works?

**Comment by dudhia** [comment]
Yes, go ahead.

**Pull request merged**